analyzers/src/SonarAnalyzer.Common/SyntaxTrackers/SyntaxTrackerBase.cs

```diff
@@ -18,35 +18,57 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
 
 namespace SonarAnalyzer.Helpers
 {
-    public abstract class SyntaxTrackerBase<TSyntaxKind>
+    public delegate bool TrackingCondition<in TContext>(TContext trackingContext);
+
+    public abstract class SyntaxTrackerBase<TSyntaxKind, TContext> : TrackerBase
         where TSyntaxKind : struct
+        where TContext : SyntaxBaseContext
     {
-        private readonly IAnalyzerConfiguration analyzerConfiguration;
-
-        protected DiagnosticDescriptor Rule { get; }
-
         protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
-
         protected abstract TSyntaxKind[] TrackedSyntaxKinds { get; }
+        protected abstract SyntaxBaseContext CreateContext(SyntaxNode expression, SemanticModel semanticModel);
 
-        protected SyntaxTrackerBase(IAnalyzerConfiguration analyzerConfiguration,
-            DiagnosticDescriptor rule)
-        {
-            this.analyzerConfiguration = analyzerConfiguration;
-            this.Rule = rule;
-        }
+        protected SyntaxTrackerBase(IAnalyzerConfiguration analyzerConfiguration, DiagnosticDescriptor rule) : base(analyzerConfiguration, rule) { }
+
+        public void Track(SonarAnalysisContext context, params TrackingCondition<TContext>[] conditions) =>
+            Track(context, new object[0], conditions);
 
-        protected bool IsEnabled(AnalyzerOptions options)
+        public void Track(SonarAnalysisContext context, object[] diagnosticMessageArgs, params TrackingCondition<TContext>[] conditions)
         {
-            analyzerConfiguration.Initialize(options);
+            context.RegisterCompilationStartAction(
+              c =>
+              {
+                  if (IsEnabled(c.Options))
+                  {
+                      c.RegisterSyntaxNodeActionInNonGenerated(
+                          GeneratedCodeRecognizer,
+                          TrackAndReportIfNecessary,
+                          TrackedSyntaxKinds);
+                  }
+              });
 
-            return analyzerConfiguration.IsEnabled(Rule.Id);
+            void TrackAndReportIfNecessary(SyntaxNodeAnalysisContext c)
+            {
+                if (CreateContext(c.Node, c.SemanticModel) is { } trackingContext
```

Comment:
What about changing the signature to this? All the way through all context constructors as well all the way down to `SyntaxContextBase`. 
```suggestion
                if (CreateContext(c) is { } trackingContext
```

Reply:
I meant to update `public SyntaxBaseContext(SyntaxNode node, SemanticModel semanticModel)` constructor to single argument as well to deduplicate the propagation of the pair Node&SemanticModel.